int32 hash
float64 x
float64 y
float64 z